### DISCUSSION TITLE
Full Stack Components no longer possible with Vite

### DISCUSSION THREAD

**Discussion by aniforprez** [remix-run/remix#8283] [Q&A] [answered]
In the latest Remix v2.4.0, a breaking change was introduced to the vite plugin (which is totally fair considering it's unstable) that route modules can no longer export arbitrary user-defined exports (See here: https://remix.run/docs/en/main/future/vite#strict-route-exports). While this definitely improves stability, this seems like it might break one of the best features about Remix which was the ability to co-locate individual UI components with their route actions (See this article by Kent Dodds: https://www.epicweb.dev/full-stack-components). I use this in my own workflow but I haven't tried out the vite plugin completely myself yet so I'm going based off the patch notes here. What is an alternative for people who still want to use this workflow and have type safety? Is it good enough that you move your component code to your components directory and do the following?

```js
import type {loader} from "~/routes/resource/thing";

export default function MyComponent(props) {
  const fetcher = useFetcher<typeof loader>()
  ...componentCode
}
```

**Comment by sergiodxa** [comment, marked as answer]
You could move the code to outside routes and then re-export the loader and action from an actual route.

**Reply by aniforprez** [reply]
Oh yes this makes a lot of sense. So I assume you could have

```js
// app/components/doTheThing.tsx

export const actionFunction = ({ request }: ActionFunctionArgs) => {
   ...doingThing
}

export function ComponentDoingThing() {
   ...doComponentThings
}
```

and in your route file have

```js
// app/routes/do-thing.ts

import {actionFunction} from "~/components/doTheThing"

export const action = actionFunction
```

**Reply by aniforprez** [reply]
I'm assuming Vite would be smart enough to know to tree shake the server and client code when deciding what to import but I'm not sure how Vite's tree shaking works. If this isn't the case, then we'd still need to split the server and client code.

**Reply by sergiodxa** [reply]
I also think it should be smart enough, but in case it’s not you will need to split them yes.

**Comment by THOUSAND-SKY** [comment]
Looks like there's a PR to remove this restriction now https://github.com/remix-run/remix/pull/8420
